File: leetcode/146-lru-cache.js
// https://leetcode.com/problems/lru-cache/ (top-interview-questions)

// Definition for doubly-linked list.
function ListNode(val) {
  this.key = key;
  this.val = val;
  this.next = null;
  this.prev = null;
}

/**
 * @param {number} capacity
 */
var LRUCache = function(capacity) {
  this.capacity = capacity;
  this.head = new ListNode(null, null);
  this.tail = new ListNode(null, null);
  this.head.next = this.tail;
  this.tail.prev = this.head;
  this.cache = new Map();

  // helper functions... append a new node to a linked list
  this.appendNode = (node) => {
    const prev = this.tail.prev;
    prev.next = node;
    node.prev = prev;
    node.next = this.tail;
    this.tail.prev = node;
  }

  // helper functions... remove a node from a doubly linked list
  this.removeNode = (node) => {
    const prev = node.prev;
    const next = node.next;
    prev.next = next;
    next.prev = prev;
  }

};

/** 
 * @param {number} key
 * @return {number}
 */
LRUCache.prototype.get = function(key) {

  if (this.cache.has(key)) {
    const node = this.cache.get(key);
    this.removeNode(node);
    this.appendNode(node);
    return node.val;
  }
  
  return -1;
};

/** 
 * @param {number} key 
 * @param {number} value
 * @return {void}
 */
LRUCache.prototype.put = function(key, value) {
  
  // in case that we have add a new key-value and we have reach our capacity...
  if (!this.cache.has(key) && this.cache.size === this.capacity) {
    // the number of keys will exceeds the capacity, so evict the least recently used key
    const lRUNode = this.head.next;
    this.cache.delete(lRUNode.key); // delete from the hash map
    this.removeNode(lRUNode); // delete from the doubly linked list
  }

  // in case that we have to replace an existed key with a new value
  if (this.cache.has(key)) {
    // delete the old one
    const node = this.cache.get(key);
    this.removeNode(node); // delete from the doubly linked list
  }


  // in the two cases we have to insert a new key - value pair
  const newNode = new ListNode(key, value);
  this.cache.set(key, newNode); // insertion in the hash map
  this.appendNode(newNode); // insertion in the doubly linked list

};

File: leetcode/146-lru-cache.py
# https://leetcode.com/problems/lru-cache/ (top-interview-questions)

class Node:
    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = dict() # key maped to: Node(key, val)
        self.head = Node(None, None) # head is a dummy node
        self.tail = Node(None, None) # tail is a dummy node
        # connect the doubly linked list head and tail together
        self.head.next = self.tail # head.next is the LRU
        self.tail.prev = self.head # tail.next is the MRU

    def get(self, key: int) -> int:
        if key in self.cache:
            node = self.cache[key]
            # update the node to be the most recently used node
            self._remove(node)
            self._append(node)
            return node.val
        return -1

    def put(self, key: int, value: int) -> None:

        if key not in self.cache and len(self.cache) == self.capacity:
            # the number of keys will exceeds the capacity
            # so evict the least recently used key
            lru_node = self.head.next
            self._remove(lru_node)
            del self.cache[lru_node.key]

        if key in self.cache:
            self._remove(self.cache[key])
        
        node = Node(key, value)
        self._append(node)
        self.cache[key] = node

    
    def _append(self, node):
        """ insert a node as the penultimate node (most recently used) """
        prev, tail = self.tail.prev, self.tail
        prev.next = node
        node.prev = prev
        tail.prev = node
        node.next = tail
    
    def _remove(self, node):
        """ remove a given node from the doubly linked list """
        prev = node.prev
        next = node.next
        prev.next = next
        next.prev = prev


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
